### DISCUSSION TITLE
Allow deep equality for collections in records

### DISCUSSION THREAD

**Discussion by dgolub** [dotnet/csharplang#5767] [Language Ideas]
Records are a great addition to the language and have the potential to save developers some time and avoid bugs by not having to write boilerplate code for things like equality.  However, the behavior of the `Equals` method on collection classes doesn't mesh well with what many developers would want for their records, which is deep equality.  Right now, if you want deep equality, you have to manually implement `Equals` and call `SequenceEqual` from LINQ on the collections in the record, which wastes the developer's time and has the potential to lead to bugs if they add a field and forget to add it to their implementation of `Equals`.  I understand that the behavior on existing collection classes can't be changed due to the need to maintain backward compatibility.  However, I see two ways that this issue could be addressed:

1. Add a new collection class called something like `RecordList` or `DeepEqualList` that override `Equals` to do deep equality.  Similarly for `DeepEqualDictionary`, `DeepEqualSet`, etc.
2. Add an attribute called something like `[DeepEqual]` or `[SequenceEqual]` that can be applied to a field in a record (or even an entire record) to signal the compiler to generate code to call `SequenceEqual` on that field (or all fields of types implementing `IEnumerable`) in the implementation of `Equals` for the record.

**Comment by TahirAhmadov** [comment]
This is a great suggestion, especially your option 1 - although it probably belongs in runtime repo. In the meantime, you should be able to make your own collection types? I don't recall, are things like `List<T>.Equals` sealed? Hopefully not, in which case it's a simple child class; otherwise you can make a wrapper or a brand new implementation (both of these have downsides, obviously).

**Comment by CyrusNajmabadi** [comment]
> Add an attribute called something like [DeepEqual] or [SequenceEqual] that can be applied to a field in a record (or even an entire record) to signal

Once at this point, you basically have a source generator.  So the guidance would be: you can do this today, without any need for language change.  Just package up these specialized equality constructs into their own analyzer/generator lib and you'll be good to go.

**Comment by huoyaoyuan** [comment]
- Collections are owned by the runtime, not the language.
- The definition of deep equality of collections is furthermore unclear. Sequence equality or set equality? Is duplicated items encountered?
- I don't even understand why to use `Equals` method.

**Reply by TahirAhmadov** [reply]
Those concerns are exaggerated, I think. The default collection equality for things like arrays and `List<T>` should be sequence equality. For `HashSet<T>` and `Dictionary<K,V>`, it should be set equality; also, for these types, duplicates aren't possible.
I hear this criticism sometimes from Java developers who are under the impression that .NET "hates" equality and it boils down to missing `Equals` overrides on `List<T>`, etc. 
The flip side is, I don't like this concept of "all objects are equatable" to begin with; however, collections are equatable if their items are. And records are not "all objects" - they are explicitly equatable.

**Reply by CyrusNajmabadi** [reply]
> For HashSet<T> and Dictionary<K,V>, it should be set equality; also

How would the language know something is a set or not?  I'm not a fan of codifying particular BCL apis here. This seems easily solvable *today* with a SG without needing any language change whatsoever.  Then you can have whatever knobs you want for your equality semantics.

**Reply by dgolub** [reply]
Both of the proposed approaches could be implemented today as a third-party library.  However, it would be desirable for something like this to be built into the language or standard library given how general-purpose it is.

**Reply by TahirAhmadov** [reply]
@CyrusNajmabadi oh yes, I don't mean the language should be changed at all. This belongs in runtime repo, not C#.

**Reply by jnm2** [reply]
@CyrusNajmabadi Knowing what it means for `ISet<T>` to be implemented by a collection doesn't feel entirely different to me from the kind of move C# makes when automatically recognizing `.RemoveAt(^1)`. I know these are different things, but it feels like the same genre of "obvious" quality of life improvements.

**Reply by CyrusNajmabadi** [reply]
i think the difference there is 1:1 relation between that syntax and a specific type.  In the discussion above it's finding a class of types.  For example, would IReadOnlyList be sequence equality?  Do we just go and find all the interfaces today in all the system collection namespaces and codify them?  I'd really like to not ever do that.

**Reply by jnm2** [reply]
@CyrusNajmabadi The only types you'd need to know about are either `IEnumerable<T>` (or `IReadOnlyCollection<T>`), `ISet<T>`, and `IReadOnlyDictionary<T>`. Maybe the mutable versions also.

**Reply by CyrusNajmabadi** [reply]
I'm not sure i agree with that.  if we look at IEnumerable for example, that presumably means we'd be doing equality *through* IEnumerable.  But i don't think that's what people would actually want.  For example, if a user has two ImmutableArrays, and the want some sort of value equality between those, it seems unlikely they would want 4 allocations of the IEnumerable/IEnumerators, as well as all teh virtual dispatches to go through on all the elements.

**Reply by HaloFour** [reply]
Not to mention the compiler should never make any assumptions as to how an `IEnumerable<T>` behaves.  Could represent a deferred operation that would perform some operation on every iteration, could be infinite, etc.

**Reply by TahirAhmadov** [reply]
I highly doubt that the language can make this work, even if it wanted to, by looking at the member types like `IEnumerable<T>`, due to all of the reasons mentioned above. The only real solution is a library one: either BCL or custom types which inherit from `List<T>` etc. and implement `Equals` appropriately.

**Reply by TahirAhmadov** [reply]
PS. Realistically, the language _could_ spit out code which tries casting the aforementioned `IEnumerable<T>` to things like `IList<T>`, etc., and perform certain pre-defined equality on those. However, that logic is extremely complicated and probably doesn't belong in the language.

**Reply by jnm2** [reply]
@CyrusNajmabadi In terms of performance it could be no different than how `foreach` lowers, and that might be seen as a plus. The presence of `IEnumerable<T>` or `IReadOnlyCollection<T>` would simply be markers to say "use sequence equality."

@ZacharyPatten Remember, I'm not arguing that this is a good thing. Only that it's not so far out of the realm of what C# does in other cases.

**Reply by ZacharyPatten** [reply]
> @ZacharyPatten Remember, I'm not arguing that this is a good thing. Only that it's not so far out of the realm of what C# does in other cases

I thumbs downed it because there would need to be a lot more changes than just the ones you mentioned.

**Reply by jnm2** [reply]
@ZacharyPatten Can you elaborate on what changes would be needed in order for the compiler to decide whether or not to use sequence equality, set equality, or dictionary equality

The question I was answering didn't have anything to do with impl strategies:

> How would the language know something is a set or not?

**Comment by sharwell** [comment]
Here's the approach I took:
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/pull/3452/commits/253d408d44f877cf8661c8bb1f84bb58b20cffa6

* Change the record parameter from `T` to `EquatableValue<T>`
* Construct the `EquatableValue<T>` with a custom equality comparer
* Expose the underlying `T` via an expression-bodied property

**Comment by alexandrehtrb** [comment]
Duplicated https://github.com/dotnet/csharplang/discussions/4845

**Reply by RikkiGibson** [reply]
I see this as a distinct request. The linked request is asking for a change in the default behavior, while this request is for something in the platform to facilitate an alternative deep equality behavior.

In my opinion, the use case for immutable, cycle-free, deep-value-equatable records is a valuable one, so I am happy to see any improvements to the experience folks can achieve, through source generation, wrapper types, or whatever else.

**Reply by alexandrehtrb** [reply]
Maybe we could have a `[DeepEquals]` attribute over records, that would make Equals and GetHashCode with structural equality over collections.

**Reply by CyrusNajmabadi** [reply]
> Maybe we could have a [DeepEquals] attribute over records, that would make Equals and GetHashCode with structural equality over collections.

This would already be possible that, without a need for anything from the language or Roslyn.  You just need to write a source generator here that looks for this attribute and which generates the right equals/gethashcode methods.

**Reply by fabianoliver** [reply]
It has been mentioned, but that is exactly what [Generator.Equals](https://github.com/diegofrata/Generator.Equals) does. Personally, I've been using this one for a while now and am rather happy with that.

I'd of course love to have sensible structural collection equality for records out of the box, but given the different possible flavors (ordered equality, unordered equality, set equality, etc etc) I'm not sure something like that could even exist so an attribute-driven way might be the best thing we can achieve anyway!
